### PR TITLE
EVG-20015: Implement new log service, log iterator, and log reader

### PR DESCRIPTION
**Pull request by julianedwards**
EVG-20015

### Description
- Implement the v0 log service that breaks a logical log into chunks and stores them as a set of files in bucket storage—the service implementation itself is extendable and since we are versioning, we can easily modify how or where we store logs with minimal impact on other Evergreen components.
- Implement two log iterators, one for chunked logs in pail-backed storage and another for merging log streams.
  - The chunk iterator code from Cedar and Logkeeper has been refactored to be simpler and lighter-weight, using a worker goroutine and a channel to fetch and consume the log chunks.
  - I also got rid of the reverse iterator and reader for reading the tail of logs—it is not actually necessary to read the logs backwards to get the last N lines, we can just skip to the last N lines and read forwards.
- Implement a log iterator reader to read and format logs from a log iterator.

### Testing
Added unit tests.

### Documentation
N/A

### CODE REVIEW

**Review by ybrill — COMMENTED** [review]
This PR triggers my imposter syndrome.

File: model/log/chunk_iterator.go [hunk 1]
```diff
@@ -0,0 +1,225 @@
...
+				// If the next channel is closed and there are
+				// no errors, this means that the worker go
+				// routine successfully exhausted all of the
+				// chunks.
+				it.exhausted = !it.catcher.HasErrors()
+				return false
+			}
+		}
+
+		data, err := it.reader.ReadString('\n')
```
[the remark this answers is not in the document]
Comment (julianedwards) [reply model/log/chunk_iterator.go:93]: Yes, exactly. The sender implementation will take care of this in the same way that the cedar buildlogger sender does.

File: model/log/chunk_iterator.go [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+	"github.com/evergreen-ci/pail"
+	"github.com/mongodb/grip"
+	"github.com/mongodb/grip/recovery"
+	"github.com/pkg/errors"
+)
+
+type chunkInfo struct {
+	prefix   string
+	key      string
+	numLines int
```
[the remark this answers is not in the document]
Comment (julianedwards) [reply model/log/chunk_iterator.go:17]: Chunks are contained to ~10MB, in any case `int` on our app servers is 64 bits.

File: model/log/merging_iterator.go [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+
+func (i *mergingIterator) Exhausted() bool {
+	exhaustedCount := 0
+	for _, it := range i.iterators {
+		if it.Exhausted() {
+			exhaustedCount++
+		}
+	}
+
+	return exhaustedCount == len(i.iterators)
```
[the remark this answers is not in the document]
Comment (julianedwards) [reply model/log/merging_iterator.go:59]: Good catch!

File: model/log/service_v0.go [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+		key:      key,
+		numLines: numLines,
+		start:    start,
+		end:      end,
+	}, nil
+}
+
+// formatRawLing returns a log line in the raw storage format.
+func (s *logServiceV0) formatRawLine(line LogLine) string {
+	return fmt.Sprintf("%d %d %s", line.Priority, line.Timestamp, line.Data)
```
[the remark this answers is not in the document]
Comment (julianedwards) [reply model/log/service_v0.go:138]: What would the benefit be there? Space? These are just the S3 keys so it doesn't really matter.

File: model/log/service_v0.go [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+		return nil, errors.Wrap(err, "iterating log chunks")
+	}
+
+	for _, chunkGroup := range chunkGroups {
+		sort.Slice(chunkGroup, func(i, j int) bool {
+			return chunkGroup[i].start > chunkGroup[j].start
+		})
+	}
+
+	return chunkGroups, nil
```
[the remark this answers is not in the document]
Comment (julianedwards) [reply model/log/service_v0.go:91]: I fixed this up—we actually do need the map to keep track of the log names.

File: model/log/reader.go [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+func NewLogIteratorReader(it LogIterator, opts LogIteratorReaderOptions) *LogIteratorReader {
+	return &LogIteratorReader{
+		it:   it,
+		opts: opts,
+	}
+}
+
+// NextTimestamp returns the Unix nanosecond timestamp of the next unread line,
+// if applicable. This should be called after the reader is exhausted,
+// successfully returning an io.EOF error. Useful for pagination.
```
[the remark this answers is not in the document]
Comment (julianedwards) [reply model/log/reader.go:53]: If we are doing byte-based pagination the underlying iterator may not be exhausted in which case `r.it.Item()` will tell us the timestamp of the next line (where we want to pick up from when we ask for the next page). I can improve the documentation to make this more clear.

File: model/log/service_v0.go [hunk 7]
```diff
@@ -0,0 +1,193 @@
...
+	return newMergingIterator(its...), nil
+}
+
+// getLogChunks maps each logical log to its chunk files stored in pail-backed
+// bucket storage for the given prefix.
+func (s *logServiceV0) getLogChunks(ctx context.Context, logNames []string) (map[string][]chunkInfo, error) {
+	logChunks := map[string][]chunkInfo{}
+
+	// To reduce potentially expensive list calls, use the LCP of the
+	// given log names when calling `bucket.List`. Key names that not have
```
Comment (ybrill) [inline model/log/service_v0.go:72]: Key names that do not have

File: go.mod [hunk 8]
```diff
@@ -140,6 +140,7 @@ require (
 	github.com/grpc-ecosystem/grpc-gateway/v2 v2.12.0 // indirect
 	github.com/hashicorp/golang-lru/v2 v2.0.1 // indirect
 	github.com/jmespath/go-jmespath v0.4.0 // indirect
+	github.com/jpillora/longestcommon v0.0.0-20161227235612-adb9d91ee629 // indirect
```
Comment (ybrill) [inline go.mod:143]: Whoever wrote this must have passed their coding interview.

**Review by ybrill — APPROVED** [review]

File: model/log/service_v0.go (hunk 7 above)
Comment (julianedwards) [reply model/log/service_v0.go]: Good catch!

**Pull request merged**